src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh

```diff
@@ -25,28 +25,37 @@ cat /tmp/toolstate/toolstates.json
 python3 "$X_PY" test --stage 2 check-tools
 python3 "$X_PY" test --stage 2 src/tools/clippy
 python3 "$X_PY" test --stage 2 src/tools/rustfmt
-python3 "$X_PY" test --stage 2 src/tools/miri
-# We natively run this script on x86_64-unknown-linux-gnu and x86_64-pc-windows-msvc.
-# Also cover some other targets via cross-testing, in particular all tier 1 targets.
+
+# Testing Miri is a bit more complicated because we want to test on multiple targets, but also try to be
+# a bit strategic so that we don't blow up our CI time.
+# We natively run this script on x86_64-unknown-linux-gnu and x86_64-pc-windows-msvc, and also cover some
+# other targets via cross-testing, in particular all tier 1 targets.
 export BOOTSTRAP_SKIP_TARGET_SANITY=1 # we don't need `cc` for these targets
 case $HOST_TARGET in
   x86_64-unknown-linux-gnu)
     # Only this branch runs in PR CI.
-    # Fully test all main OSes, including a 32bit target.
-    python3 "$X_PY" test --stage 2 src/tools/miri --target x86_64-apple-darwin
-    python3 "$X_PY" test --stage 2 src/tools/miri --target i686-pc-windows-msvc
-    # Only run "pass" tests for the remaining targets, which is quite a bit faster.
-    python3 "$X_PY" test --stage 2 src/tools/miri --target x86_64-pc-windows-gnu --test-args pass
-    python3 "$X_PY" test --stage 2 src/tools/miri --target i686-unknown-linux-gnu --test-args pass
-    python3 "$X_PY" test --stage 2 src/tools/miri --target aarch64-unknown-linux-gnu --test-args pass
-    python3 "$X_PY" test --stage 2 src/tools/miri --target s390x-unknown-linux-gnu --test-args pass
+    # We set the GC interval to the shortest possible value (0 would be off) to increase the chance
+    # that bugs which only surface when the GC runs at a specific time are more likely to cause CI to fail.
+    # But this significantly increases the runtime of the test suite, so we just do it for one target.
+    # https://github.com/rust-lang/rust/pull/118029#issuecomment-1817751089
+    MIRIFLAGS=-Zmiri-provenance-gc=1 python3 "$X_PY" test --stage 2 src/tools/miri
```

Comment:
So this is the only actual change, everything else is refactoring this script?

Reply:
Yeah, that's what I eventually landed on :shrug: I'm not really sure where to get the CI time budget that we need to run more tests. I'll probably try some other tweaks to which tests we run in a later PR, but I don't want to block this one.

Reply:
I'm not even sure what the budget is, t-infra manages that. This runner is a fast one so making it a bit slower won't slow down bors. However we might want to differentiate "running in PR CI" and "running in bors check".

Reply:
Hm I was only thinking of running this in PR CI. I'll see if I can ramp up the tests for bors.

Reply:
Honestly I think having the provenance-gc on one target is completely fine here. I meant to suggest to only do that with bors, and revert back to the old behavior for PR CI.

Reply:
That seems reasonable.

By "budget" I was mostly referring to not adding a lot of time to people's PR CI runs. `-tools` is already the slowest job or very nearly the slowest, so I'd prefer to not make it any more slow than it needs to be.